src/PluginInterface.ts

```diff
@@ -21,9 +21,14 @@ export default class PluginInterface<T extends CompilerPlugin = CompilerPlugin>
         for (let plugin of this.plugins) {
             if ((plugin as any)[event]) {
                 try {
-                    this.logger.time(LogLevel.debug, [plugin.name, event], () => {
-                        (plugin as any)[event](...args);
+                    const returnValue = this.logger.time(LogLevel.debug, [plugin.name, event], () => {
+                        return (plugin as any)[event](...args);
                     });
+
+                    //plugins can short-circuit the event by returning `false`
+                    if (returnValue === false) {
```

Comment:
Should that be the opposite - return `true` if captured?

Reply:
I following the design of the browser event model, where returning false cancels the entire chain. https://stackoverflow.com/a/30473685/1633757
I think of it as `"shouldContinue?: true|false"`. 

I'm open to switching it still if you feel strongly about it, but `false` makes more sense to me.

Reply:
@georgejecook @triwav @chrisdp care to weigh in on this?

Reply:
This is what I'd expect, too (false to short-circuit), for this kind of mechanism; but what are we cancelling here? It looks like we are preventing any other plugins getting events here, right?

Reply:
I think for `onHover` it's not about cancelling but handling, as when we confirm a key event is handled.

Reply:
That said ultimately I don't have a strong opinion on either options.

Reply:
when I read this, it looks like if I return false, then no other plugin will be notified of this event, as we exit the function, so I'm not reading this as opt-in to handle event; but return false, and no more plugins events.

![image](https://user-images.githubusercontent.com/1214545/115853241-400adc00-a429-11eb-9502-0508bd62b6a6.png)

imo: If it's how I read it, it _should_ be false, if it's opt-in, like @elsassph says, it should be true

Reply:
I'm not a fan of this though; as I could potentiallly have to consdier other people's plugins, when debugging my own. what's the use case?

Reply:
The use case is in the description of this pr: 
>This is useful for onGetHover because only a single plugin may contribute a hover, so the first one that contributes needs to indicate they did so.

For `onGetCompletions`, every plugin may contribute (add to) the result...it's cumulative. However, for `onGetHover`, the language server protocol only supports a single `hover` result. Which means we should stop looking for a hover result the moment we find one. that's what this is for. the first plugin who responds to this, wins.

Reply:
I think most of my thinking comes from the fact that you return true on Roku to say you handled something and it shouldn't continue up the node tree. One other idea. This would take more resources but if instead of exiting out on the first you let all the plugins go. Each one can return a priority on it and the highest one wins. Not huge fan of it as if we ever have community plugins there'd need to be an agreed upon priority scale for it.

Reply:
Yes @triwav that's also what I'm thinking here - it's not a "cancellation" pattern.
We want plugins to say, for some events, "I have handled the event, stop propagating" like Roku key events.
And to avoid weird issues I'd also suggest having a separte `emitUntilHandled` method.

Reply:
Given that, I agree @elsassph, I'd expect to return true for this.

Reply:
I entirely refactored this logic so that plugins can all contribute one or more hovers, and we merge all the text together into a single response to the client. In this way, we no longer need to short-circuit at all. A plugin can decide to contribute or not based on whether the `event.hovers` array is empty.